### PR TITLE
[14.0] Migrate queue_job_cron

### PR DESCRIPTION
**Pull request by codingforfun**

### CODE REVIEW

**Comment by fshah-initos** [comment]
@all can someone please review this PR?

File: queue_job_cron/models/ir_cron.py [hunk 1]
```diff
@@ -32,14 +32,19 @@ def _run_job_as_queue_job(self, server_action):
...
-        if self.run_as_queue_job:
-            return self.with_delay(
-                priority=self.priority,
-                description=self.name,
-                channel=self.channel_id.complete_name,
-            )._run_job_as_queue_job(server_action=self.ir_actions_server_id)
-        else:
-            return super().method_direct_trigger()
+        for cron in self:
+            if not self.run_as_queue_job:
```
Comment (simahawk) [inline queue_job_cron/models/ir_cron.py]:
```suggestion
            if not cron.run_as_queue_job:
```

File: queue_job_cron/models/ir_cron.py [hunk 2]
```diff
@@ -32,14 +32,19 @@ def _run_job_as_queue_job(self, server_action):
...
+                super(IrCron, cron).method_direct_trigger()
```
Comment (simahawk) [inline queue_job_cron/models/ir_cron.py:37]:
```suggestion
                super().method_direct_trigger()
```

File: queue_job/models/queue_job.py [hunk 3]
```diff
@@ -296,7 +296,6 @@ def autovacuum(self):
                 )
                 if jobs:
                     jobs.unlink()
-                    self.env.cr.commit()
```
Comment (simahawk) [inline queue_job/models/queue_job.py:299]: can you explain this change?

**Review by simahawk — CHANGES_REQUESTED** [review]

**Comment by simahawk** [comment]
> @ALL can someone please review this PR?

where is yours? :wink:

File: queue_job_cron/models/ir_cron.py (hunk 1 above)
Comment (codingforfun) [reply queue_job_cron/models/ir_cron.py]: You’r right. Thx.

File: queue_job/models/queue_job.py (hunk 3 above)
Comment (codingforfun) [reply queue_job/models/queue_job.py:299]: Hmm. I don’t see this change. It’s not part of my patch.

**Comment by codingforfun** [comment]
> > @ALL can someone please review this PR?
> 
> where is yours? wink

I don’t know if I understand the question right, but I think you are asking who is responsible here. This is we have different roles in the team here. I made the PR because I needed it urgently for a project, but my co-workers are looking at a bunch of OCA modules for migration to 14.0 anyways so I asked them to take care, respond to question etc.

File: queue_job/models/queue_job.py (hunk 3 above)
Comment (codingforfun) [reply queue_job/models/queue_job.py:299]: I checked in `git blame`. This comes from 085d5ba8.

**Comment by codingforfun** [comment]
pre-commit fails because of a change done in 085d5ba8

**Comment by simahawk** [comment]
> > > @ALL can someone please review this PR?
> > 
> > 
> > where is yours? wink
> 
> I don’t know if I understand the question right, but I think you are asking who is responsible here. This is we have different roles in the team here. I made the PR because I needed it urgently for a project, but my co-workers are looking at a bunch of OCA modules for migration to 14.0 anyways so I asked them to take care, respond to question etc.

Indeed it was for @fshah-initos. I was trying to encourage him to drop a review :wink:

File: queue_job_cron/models/ir_cron.py [hunk 4]
```diff
@@ -32,14 +32,17 @@ def _run_job_as_queue_job(self, server_action):
...
-            return self.with_delay(
-                priority=self.priority,
-                description=self.name,
-                channel=self.channel_id.complete_name,
-            )._run_job_as_queue_job(server_action=self.ir_actions_server_id)
-        else:
-            return super().method_direct_trigger()
+        for cron in self:
+            if not cron.run_as_queue_job:
+                super().method_direct_trigger()
```
Comment (simahawk) [inline queue_job_cron/models/ir_cron.py]: the call to super here should be done using cron instead of self

Comment (codingforfun) [reply queue_job_cron/models/ir_cron.py]: You mean like this? 
```
super(IrCron, cron).method_direct_trigger()
``` 
Funny. You proposed the opposite thing yesterday and I was thinking about this to short and didn’t remember why I did it like this, so just took your suggestion. ;-)

Comment (codingforfun) [reply queue_job_cron/models/ir_cron.py]: Thx anyway.

Comment (simahawk) [reply queue_job_cron/models/ir_cron.py]: eh, sorry, I've deleted the comment right after as I realized it was correct :stuck_out_tongue_winking_eye: 
Maybe you still had it in your notification :man_shrugging:

File: queue_job_cron/models/ir_cron.py [hunk 5]
```diff
@@ -32,14 +32,17 @@ def _run_job_as_queue_job(self, server_action):
...
-        else:
-            return super().method_direct_trigger()
+        for cron in self:
+            if not cron.run_as_queue_job:
+                super(IrCron, cron).method_direct_trigger()
+            else:
+                _cron = cron.with_user(cron.user_id).with_context(lastcall=cron.lastcall)
+                _cron.with_delay(
+                    priority=self.priority,
+                    description=self.name,
```
Comment (simahawk) [inline queue_job_cron/models/ir_cron.py]: but then... all the `self` here should be replaced w/ `cron` :wink:

Comment (simahawk) [reply queue_job_cron/models/ir_cron.py]: tip: next time delegate this to a method `_run_as_queue_job()` and forget about this mess

Comment (codingforfun) [reply queue_job_cron/models/ir_cron.py]: Yes. You see the quick hack nature of this.

**Comment by fshah-initos** [comment]
Hi @simahawk 
Can you please review the PR?

File: queue_job_cron/models/ir_cron.py [hunk 6]
```diff
@@ -32,14 +32,19 @@ def _run_job_as_queue_job(self, server_action):
...
+            if not cron.run_as_queue_job:
+                super(IrCron, cron).method_direct_trigger()
+            else:
+                _cron = cron.with_user(cron.user_id).with_context(
+                    lastcall=cron.lastcall
+                )
+                _cron.with_delay(
+                    priority=self.priority,
+                    description=self.name,
+                    channel=self.channel_id.complete_name,
```
Comment (simahawk) [inline queue_job_cron/models/ir_cron.py:45]:
```suggestion
                    channel=cron.channel_id.complete_name,
```
no?

Comment (simahawk) [inline queue_job_cron/models/ir_cron.py]:
```suggestion
                    priority=_cron.priority,
                    description=_cron.name,
                    channel=_cron.channel_id.complete_name,
```

**Review by simahawk — CHANGES_REQUESTED** [review]

**Comment by fshah-initos** [comment]
@simahawk I have implemented the changes suggested. Will you please review it?

**Comment by simahawk** [comment]
@fshah-initos good, thank you.
One last step: squash the last 2 commits and re-word the 1st one as `[MIG] queue_job_cron: Migration to 13.0`:

**Comment by fshah-initos** [comment]
Hi @simahawk 
I have squash the commit and re-word it. Can this be merged?

**Review by simahawk — APPROVED** [review]

**Comment by simahawk** [comment]
> Hi @simahawk
> I have squash the commit and re-word it. Can this be merged?

maybe you can re-review and add your review? :)

**Review by fshah-initos — APPROVED** [review]

**Comment by fshah-initos** [comment]
@all Is it possible to merge this PR?

**Comment by fshah-initos** [comment]
Can someone from the maintainers please merge the PR?

**Comment by fshah-initos** [comment]
It would be great if somone could merge this PR.

**Comment by simahawk** [comment]
@fshah-initos this comment is not attended https://github.com/OCA/queue/pull/315#issuecomment-804745259 
Not blocking now but remember it next time :wink: 

/ocabot merge nobump

**Pull request merged**